### PR TITLE
More sophisticated measurement validation for default-qubit

### PR DESCRIPTION
**Pull request by astralcai**
**Description of the Change:**
Add measurement-specific observable validations to measurement validation

[sc-64700]

### CODE REVIEW

File: pennylane/devices/default_qubit.py [hunk 1]
```diff
@@ -103,16 +79,74 @@ def stopping_condition_shots(op: qml.operation.Operator) -> bool:
...
+    if isinstance(obs, qml.ops.Hamiltonian):
+        return all(observable_accepts_sampling(o) for o in obs.ops)
+
+    if isinstance(obs, qml.operation.Tensor):
+        return all(observable_accepts_sampling(o) for o in obs.obs)
+
+    return obs.has_diagonalizing_gates
+
+
+def observable_accepts_analytic(obs: qml.operation.Operator, is_expval=False) -> bool:
```
Comment (mudit2812) [inline pennylane/devices/default_qubit.py:100]:
```suggestion
def observable_accepts_analytic(obs: qml.operation.Observable, is_expval=False) -> bool:
```

File: pennylane/devices/default_qubit.py [hunk 2]
```diff
@@ -103,16 +79,74 @@ def stopping_condition_shots(op: qml.operation.Operator) -> bool:
     return isinstance(op, (Conditional, MidMeasureMP)) or stopping_condition(op)
 
 
+def observable_accepts_sampling(obs: qml.operation.Operator) -> bool:
```
Comment (mudit2812) [inline pennylane/devices/default_qubit.py:82]:
```suggestion
def observable_accepts_sampling(obs: qml.operation.Observable) -> bool:
```

File: pennylane/devices/preprocess.py [hunk 3]
```diff
@@ -479,7 +479,9 @@ def sample_measurements(m):
         and not isinstance(meas := op.hyperparameters["measurement"], qml.measurements.StateMP)
     ]
 
-    if tape.shots:
+    shots = qml.measurements.Shots(tape.shots)
+
+    if shots.total_shots is not None:
```
Comment (mudit2812) [inline pennylane/devices/preprocess.py:484]: I'm curious about this change. `tape.shots` should always be of type `Shots`

File: tests/workflow/test_construct_batch.py [hunk 4]
```diff
@@ -140,7 +140,7 @@ def circuit(x):
             return qml.expval(qml.PauliZ(0))
 
         full_prog = get_transform_program(circuit)
-        assert len(full_prog) == 13
+        assert len(full_prog) == 12
```
Comment (mudit2812) [inline tests/workflow/test_construct_batch.py:143]: What made the transform program smaller?

File: tests/devices/qubit/test_measure.py [hunk 5]
```diff
@@ -179,12 +179,9 @@ def test_op_math_observable_jit_compatible(self):
 
         dev = qml.device("default.qubit", wires=4)
 
-        O1 = qml.X(0)
-        O2 = qml.X(0)
-
         @qml.qnode(dev, interface="jax")
         def qnode(t1, t2):
-            return qml.expval(qml.prod(O1, qml.RX(t1, 0), O2, qml.RX(t2, 0)))
+            return qml.expval((t1 * qml.X(0)) @ (t2 * qml.Y(1)))
```
Comment (mudit2812) [inline tests/devices/qubit/test_measure.py:184]: Nice that the new validation actually works with jit :)

**Review by mudit2812 — APPROVED** [review]
Couple of non-blocking comments.

File: pennylane/devices/preprocess.py (hunk 3 above)
Comment (astralcai) [reply pennylane/devices/preprocess.py:484]: Not with the legacy device iirc

File: tests/workflow/test_construct_batch.py (hunk 4 above)
Comment (astralcai) [reply tests/workflow/test_construct_batch.py:143]: Previously there was a `validate_observables` transform, now that is removed and `validate_measurements` now validates the observables as well depending on what the measurement is

**Review by lillian542 — COMMENTED** [review]
Looks good! 🚀 

I guess moving away from a fixed set of observables to this check regarding diagonalizing gates creates the opportunity to define your own observables. Since that's sort of an 'added functionality' here, should we add a test for it?

**Review by lillian542 — APPROVED** [review]

**Pull request merged**